### PR TITLE
Use IndexOf in StringBuilder.Replace(string, string, ...)

### LINKED ISSUE #80952: Lacking performance of `StringBuilder.Replace`
### Description

When comparing a chain of `StringBuilder.Replace` calls to replace two or more values from a given input string, its performance is incredibly lacking. This is compared to chaining several `string.Replace` calls after one another.

### Configuration

```
BenchmarkDotNet=v0.13.3, OS=Windows 11 (10.0.22621.1105)
AMD Ryzen 5 PRO 4650U with Radeon Graphics, 1 CPU, 12 logical and 6 physical cores
.NET SDK=7.0.102
  [Host]   : .NET 6.0.13 (6.0.1322.58009), X64 RyuJIT AVX2
  .NET 6.0 : .NET 6.0.13 (6.0.1322.58009), X64 RyuJIT AVX2
  .NET 7.0 : .NET 7.0.2 (7.0.222.60605), X64 RyuJIT AVX2
```

### Data

|                Method |  Runtime |      Mean |    Error |   StdDev | Ratio | RatioSD |
|---------------------- |--------- |----------:|---------:|---------:|------:|--------:|
| Replace_StringReplace | .NET 6.0 | 184.60 us | 2.713 us | 2.265 us |  1.00 |    0.00 |
| Replace_StringBuilder | .NET 6.0 | 471.90 us | 0.821 us | 0.686 us |  2.56 |    0.03 |
|                       |          |           |          |          |       |         |
| Replace_StringReplace | .NET 7.0 |  22.07 us | 0.140 us | 0.124 us |  1.00 |    0.00 |
| Replace_StringBuilder | .NET 7.0 | 555.60 us | 0.756 us | 0.631 us | 25.17 |    0.16 |

Benchmark code: https://gist.github.com/just-ero/aefc9b5123b69c3ef49944fe6242a104.

### Regression

Seems as though `StringBuilder.Replace` used to be faster on .NET 6.0.

### Analysis

`string.Replace` does a vectorized IndexOf search.  
`StringBuilder.Replace` walks its `char` buffer and does `StartsWith` at each position.

Pretty obvious how much of a performance hit that is.

### PR DESCRIPTION
**Pull request by stephentoub**
StringBuilder.Replace(string, string, ...) currently walks character-by-character to find the next location of the oldValue.  This change adds the use of MemoryExtensions.IndexOf to search for the oldValue, and only falls back to the character-by-character walk when it nears and needs to account for a chunk boundary.  It also switches to track replacement indices using more stack space and the array pool (via ValueListBuilder) rather than less stack space and array allocations.

Using the benchmark from https://github.com/dotnet/runtime/issues/80952:

|                Method |         Toolchain |      Mean |    Error |   StdDev | Ratio | Allocated | Alloc Ratio |
|---------------------- |------------------ |----------:|---------:|---------:|------:|----------:|------------:|
| Replace_StringReplace | \main\corerun.exe |  43.71 us | 0.537 us | 0.476 us |  1.00 | 227.96 KB |        1.00 |
| Replace_StringBuilder | \main\corerun.exe | 507.24 us | 3.251 us | 2.882 us | 11.60 |  62.11 KB |        0.27 |
| Replace_StringReplace |   \pr\corerun.exe |  39.38 us | 0.440 us | 0.411 us |  0.90 | 227.96 KB |        1.00 |
| Replace_StringBuilder |   \pr\corerun.exe |  24.16 us | 0.213 us | 0.189 us |  0.55 |   58.9 KB |        0.26 |

Fixes https://github.com/dotnet/runtime/issues/80952

### CODE REVIEW

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 1]
```diff
@@ -1944,45 +1943,89 @@ public StringBuilder Replace(string oldValue, string? newValue, int startIndex,
...
+                        count -= move;
+                        break;
                     }
-                    replacements[replacementsCount++] = indexInChunk;
-                    indexInChunk += oldValue.Length;
-                    count -= oldValue.Length;
                 }
-                else
+
+                Debug.Assert(oldValue.Length > chunk.m_ChunkLength - indexInChunk || count == 0,
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1986]: This assert has failed in https://helixre107v0xdeko0k025g8.blob.core.windows.net/dotnet-runtime-refs-pull-81098-merge-25df720c9d4e4a9683/System.Runtime.Tests/1/console.4fb7bee6.log?helixlogtype=result:

```log
  Starting:    System.Runtime.Tests (parallel test collections = on, max threads = 2)
Process terminated. Assertion failed.
oldValue.Length = 3, chunk.m_ChunkLength - indexInChunk = 14, count == 2
   at System.Text.StringBuilder.Replace(String oldValue, String newValue, Int32 startIndex, Int32 count) in /_/src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:line 1986
   at System.Text.Tests.StringBuilderTests.Replace_String(String value, String oldValue, String newValue, Int32 startIndex, Int32 count, String expected) in /_/src/libraries/System.Runtime/tests/System/Text/StringBuilderTests.cs:line 1727
   at InvokeStub_StringBuilderTests.Replace_String(Object, Object, IntPtr*)
```

Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1986]: It's a bogus assert.  Will fix.

**Comment by stephentoub** [comment]
@adamsitnik, the assert issue is fixed.  Can you take another look? Thanks.

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 2]
```diff
@@ -1934,8 +1934,7 @@ public StringBuilder Replace(string oldValue, string? newValue, int startIndex,
 
             newValue ??= string.Empty;
 
-            Span<int> replacements = stackalloc int[5]; // A list of replacement positions in a chunk to apply
-            int replacementsCount = 0;
+            var replacements = new ValueListBuilder<int>(stackalloc int[128]); // A list of replacement positions in a chunk to apply
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1937]: great use case for `ValueListBuilder` 👍 I am suprised that the temporary array allocation survived for so long in this repo ;)

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs [hunk 3]
```diff
@@ -1944,45 +1943,89 @@ public StringBuilder Replace(string oldValue, string? newValue, int startIndex,
...
+
+                Debug.Assert(oldValue.Length > Math.Min(count, chunk.m_ChunkLength - indexInChunk),
+                    $"oldValue.Length = {oldValue.Length}, chunk.m_ChunkLength - indexInChunk = {chunk.m_ChunkLength - indexInChunk}, count == {count}");
+
+                // Now do the more complicated cross-chunk matching.
+                while (indexInChunk < chunk.m_ChunkLength && count > 0)
                 {
-                    indexInChunk++;
-                    --count;
+                    if (StartsWith(chunk, indexInChunk, count, oldValue))
```
Comment (adamsitnik) [inline src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1992]: this method is the only method using `StartsWith`, would we gain anything from inlining it? Would you mind running the benchmarks if you still have this branch built on your machine? If not, it's not worth your time (I am just curious and it's really hard to suggest any improvements for this PR)

**Review by adamsitnik — APPROVED** [review]
LGTM, thank you very much @stephentoub !

File: src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs (hunk 3 above)
Comment (stephentoub) [reply src/libraries/System.Private.CoreLib/src/System/Text/StringBuilder.cs:1992]: > if you still have this branch built on your machine

I don't.  We can follow-up on your question separately, though.

**Pull request merged**